Resolve JSON schema reference paths for relative imports

benoittgt opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

When validating values against JSON schemas, relative file references in the schema (like "$ref": "./other-schema.json") were failing in some cases. Main reason: The JSON schema validator couldn't resolve the correct base directory.

This change passes the absolute base directory to the schema validator, allowing it to properly resolve relative JSON schema references by constructing the correct file:// URLs with the proper path context.

This is another proposal of https://github.com/helm/helm/pull/31261.

Tested with:
- `helm lint .`
- `helm lint ../chart`
- `helm install ../chart`
- `helm template ../chart`
- `helm upgrade ../chart`

With tgz it will fail the same as before except for lint command that handle it differently.

Fixes #31260
Close https://github.com/helm/helm/pull/31261

**Special notes for your reviewer**:

To test it manually:
- Use a [chart](https://github.com/user-attachments/files/22252417/sample-chart-0.1.0.tgz) that reference in `value.schema.json` another file and put into a subfolder.

```js
// value.schema.json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Helm Chart Values",
  "description": "Schema for sample-chart values.yaml",
  "type": "object",
  "properties": {
    "image": {
      "$ref": "./library.json"
    }
  },
  "additionalProperties": true
}

// library.json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Container Image Configuration",
  "description": "Schema for container image configuration",
  "type": "object",
  "properties": {
    "repository": {
      "type": "string"
    },
    "pullPolicy": {
      "type": "string"
    },
    "tag": {
      "type": "string"
    }
  },
  "additionalProperties": false
}
```

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

benoittgt committed:
Resolve JSON schema reference paths for relative imports

When validating values against JSON schemas, relative file references
in the schema (like "$ref": "./other-schema.json") were failing because
in some cases. The JSON schema validator couldn't resolve the correct
base directory.

This change passes the absolute base directory to the schema validator,
allowing it to properly resolve relative JSON schema references by
constructing the correct file:// URLs with the proper path context.

Tested with:
- `helm lint .`
- `helm lint subfolder`

Fixes #31260

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

Add test

benoittgt committed:
Do not remove necessary logs

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Fix error message trimming for dynamic schema paths in JSON validation

Previously, the error trimming was hardcoded to remove only:
'jsonschema validation failed with 'file:///values.schema.json#''

After the fix for relative schema references (issue #31260), schema URLs
now include the full absolute path, making error messages look like:
'jsonschema validation failed with 'file:///full/absolute/path/values.schema.json#''

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented:
Oups sorry missed that.

benoittgt commented:
@robertsirc it should be good now.

gjenkins8 commented on pkg/chart/common/util/jsonschema.go:
```diff
@@ -148,12 +149,13 @@ func ValidateAgainstSingleSchema(values common.Values, schemaJSON []byte) (reter
 	compiler := jsonschema.NewCompiler()
 	compiler.UseLoader(loader)
-	err = compiler.AddResource("file:///values.schema.json", schema)
+	base := "file://" + filepath.ToSlash(absBaseDir) + "/values.schema.json"
```
This line here I think is the core of the fix?

But... why does specifying the absolute path of `values.schema.json`, rather than the relative path, help here?

If the loader is able to load from (presumably) the current working directory. Why wouldn't relative schema paths also be loaded relative to the current working directory?

benoittgt commented:
I will take a look. Sorry for the poor quality of the contribution.

benoittgt committed:
fix(schema): Add schema $ref resolution for all commands

Extends JSON schema validation to support relative $ref imports
across helm lint, install, template, and upgrade commands.

Addresses George's review comments:
- Pass chartDir directly instead of deriving from valuesPath
- Create new functions instead of changing public API signatures
- Use absolute file paths for proper URL-based $ref resolution

Implementation:
- Added ValidateAgainstSchemaWithPath to resolve $ref using
  absolute file paths instead of root-relative URLs
- Thread chartDir through Install/Upgrade structs and all
  validation functions
- Updated both v2 (public) and v3 (internal) lint rules

Fixes relative $ref resolution when running commands like:
  helm lint ../sample-chart
  helm install release ../sample-chart

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
fix: handle dynamic schema paths in error message trimming

The error message trimming logic was hardcoded to remove the prefix
"jsonschema validation failed with 'file:///values.schema.json#'",
which broke when we switched to using absolute paths for $ref resolution.

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
fix(test): Remove symlink

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented:
Another try. Not sure about the `ChartDir` here.

benoittgt committed:
fix(schema): limit $ref resolution to lint command

Schema validation with $ref cross-file references only works when charts
are available as directories on the filesystem. The install, template, and
upgrade commands may work with archived charts (.tgz) that are loaded
directly into memory without filesystem extraction.

This change limits ChartDir setting to the lint command only. The lint
command already extracts archived charts to temporary directories before
validation, so $ref resolution works for both directory and archived charts
when linting.

For install/template/upgrade commands:
- Main schema validation continues to work
- Charts without $ref validate successfully
- Charts with $ref will fail .. maybe we could add a warning about $ref not being supported in these commands?

Changes:
- Remove ChartDir assignment from install and upgrade commands
- Update schema validation to use synthetic path when chartDir is empty
- Add test for linting archived chart with $ref
- Remove install/template/upgrade tests for $ref (not supported)

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
fix(schema): enable $ref resolution for directory charts

Set ChartDir only for directory-based charts to enable $ref resolution in
JSON schemas. Archived charts (.tgz) are loaded into memory without filesystem
extraction, so $ref resolution is not supported for them.

This fixes the original issue where `helm template .` and `helm install .`
failed to validate schemas with relative $ref references.

Fixes #31260

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
fix(schema): enable $ref resolution for directory charts

Set ChartDir only for directory-based charts to enable $ref resolution in
JSON schemas. Archived charts (.tgz) are loaded into memory without filesystem
extraction, so $ref resolution is not supported for them.

This fixes the original issue where `helm template .` and `helm install .`
failed to validate schemas with relative $ref references.

Fixes #31260

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
fix(schema): resolve $ref for aliased subchart dependencies

After `processDependencyEnabled`, aliased subcharts have their
`Metadata.Name` rewritten to the alias, but the on-disk directory
retains the original chart name. The previous code used `sub.Name()`
directly to build the filesystem path, causing `$ref` resolution to
look in a non-existent directory (e.g., `charts/database/` instead
of `charts/mysql/`).

Add `resolveSubchartDir` which first tries a direct name match, then
falls back to scanning `charts/` subdirectories by comparing
`values.schema.json` content byte-for-byte with the in-memory schema.

Addresses https://github.com/helm/helm/pull/31274#discussion_r3224792920

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented:
I just pushed the last commits addressing the Copilot feedback https://github.com/helm/helm/pull/31274#discussion_r3224792920. Tested locally with a few scenarios:

1. When you have a chart with a `$ref` in `values.schema.json`, `helm lint` / `template` / `install --dry-run` all resolve it correctly from absolute path, cwd, and relative path.
2. When you have invalid values that violate the referenced schema's constraint, the error is caught and reported by both `lint` and `template`
3. When you have a subchart aliased in `Chart.yaml` (example `mylib` on disk, `database` as alias), `lint` / `template` / `install --dry-run` all pass. `resolveSubchartDir()` correctly falls back to bytematching `values.schema.json` to find the right directory
4. When you package the chart as a `.tgz`, `lint` still passes (it extracts to a temp dir), but `template` fails as expected since there's no filesystem path to resolve `$ref` from. 👌

benoittgt committed:
chore: Linting sprintf

Error: pkg/chart/common/util/jsonschema.go:194:15: string-format: fmt.Sprintf can be replaced with string concatenation (perfsprint)
Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>
